pkg/stacks/unpack.go

```diff
@@ -17,320 +17,439 @@ limitations under the License.
-			resourcesFiles = append(resourcesFiles, path)
-		}
-		return nil
-	})
-	if err != nil {
-		// walking for resources encountered an error that halted it, surface this error
-		return nil, err
+// NewStackPackage returns a StackPackage with maps created
+func NewStackPackage() *StackPackage {
+	// create a Stack record and populate it with the relevant package contents
+	v, k := v1alpha1.StackGroupVersionKind.ToAPIVersionAndKind()
+
+	sp := &StackPackage{
+		Stack: v1alpha1.Stack{
+			TypeMeta: metav1.TypeMeta{APIVersion: v, Kind: k},
+			ObjectMeta: metav1.ObjectMeta{
+				Annotations: map[string]string{},
+			},
+		},
+		CRDs:      map[schema.GroupVersionKind]apiextensions.CustomResourceDefinition{},
+		CRDPaths:  map[schema.GroupVersionKind]string{},
+		Groups:    map[string]StackGroup{},
+		Icons:     map[string]*v1alpha1.IconSpec{},
+		Resources: map[string]StackResource{},
+		UISpecs:   map[string]string{},
 	}
 
-	return resourcesFiles, nil
+	return sp
 }
 
-// Ignore interfacer linter check for this function signature, it wants to use io.StringWriter instead
-// of strings.Builder, but io.StringWriter is only available in go 1.12+ and we support older versions.
-// nolint:interfacer
-func readResourceFile(fs afero.Fs, rf string, crdList *v1alpha1.CRDList, sb *strings.Builder) error {
-	b, err := afero.ReadFile(fs, rf)
-	if err != nil {
-		// we weren't able to read the current resource file, surface this error
+// Unpack the stack contents from the given directory and output all Stack artifacts to stdout as YAML
+func Unpack(rw walker.ResourceWalker, out io.StringWriter) error {
+	log.V(logging.Debug).Info("Unpacking stack")
+
+	sp := NewStackPackage()
+
+	rw.AddStep(appFileName, appStep(sp))
+
+	rw.AddStep(groupFileName, groupStep(sp))
+
+	rw.AddStep(resourceFileNamePattern, resourceStep(sp))
+	rw.AddStep(crdFileNamePattern, crdStep(sp))
+	rw.AddStep(permissionsFileName, rbacStep(sp))
+	rw.AddStep(installFileName, installStep(sp))
+	rw.AddStep(iconFileNamePattern, iconStep(sp))
+	rw.AddStep(uiSpecFileNamePattern, uiStep(sp))
+
+	if err := rw.Walk(); err != nil {
 		return err
 	}
 
-	// unmarshal the raw resource file content into a CRD type
-	var crd apiextensions.CustomResourceDefinition
-	if err := yaml.Unmarshal(b, &crd); err != nil {
-		return err
+	if !sp.GotApp() {
+		// TODO(displague) is it an error to unpack an stack without a app.yaml file or an incomplete one?
+		return nil
 	}
 
-	// add the CRD type meta to the list
-	// TODO(jbw976): handle cases where the CRD has multiple versions associated with it
-	crdGVK := schema.GroupVersionKind{
-		Group:   crd.Spec.Group,
-		Version: crd.Spec.Version,
-		Kind:    crd.Spec.Names.Kind,
+	sp.applyAnnotations()
+
+	_, err := out.WriteString(sp.Yaml())
+
+	return err
+}
+
+// orderStackGroupKeys returns the map indexes in descending order (/foo/bar/baz, /foo/bar, /foo, /bar)
+func orderStackGroupKeys(m map[string]StackGroup) []string {
```

Comment:
what were you saying the other day about generics? 🤢

Reply:
I added comments about this in https://github.com/crossplaneio/crossplane/compare/9f5ab3c6d421d5e1b0d722af3317569614d5132f..017c3d47090daca29f82dfd30b95578a3b3e297d 

Let me know if you would prefer another approach.  Since the annotation functions now share the same signature I could come up with a set of interfaces around:
* getting map indexes
* ordering indexes
* annotating using index orderable mapped things

I think the harnessing might add more complexity than the code duplication, at this point.

Reply:
reflection could the duplication. https://stackoverflow.com/a/39072207/156674

Reply:
i don't strongly feel we need to change this or avoid the duplication, that's your call.  I was just bringing up that we had talked about generics recently and that their lacking in go is gross :)

Reply:
Sorry to say, I added yet another.. https://github.com/crossplaneio/crossplane/compare/fc10aac76fdf8c71257f77a6fb82fa19ad33decc..59a87bdad017210353e5fe5ff551ef5a9fbb55ce

This should be easy to clean up later, using reflection or some other pattern.

I replaced the `schema.GroupVersionKind` map keys with their string equivalents. The maps now have `map[string]T` in common.